edge/pkg/edged/volume/csi/csi_attacher.go

```diff
@@ -0,0 +1,447 @@
+	return "", nil
+}
+
+func (c *csiAttacher) WaitForAttach(spec *volume.Spec, _ string, pod *v1.Pod, timeout time.Duration) (string, error) {
+	source, err := getPVSourceFromSpec(spec)
+	if err != nil {
+		klog.Error(log("attacher.WaitForAttach failed to extract CSI volume source: %v", err))
+		return "", err
+	}
+
+	attachID := getAttachmentName(source.VolumeHandle, source.Driver, string(c.plugin.host.GetNodeName()))
+
+	return c.waitForVolumeAttachment(source.VolumeHandle, attachID, timeout)
+}
+
+func (c *csiAttacher) waitForVolumeAttachment(volumeHandle, attachID string, timeout time.Duration) (string, error) {
+	klog.V(4).Info(log("probing for updates from CSI driver for [attachment.ID=%v]", attachID))
+
+	err := wait.PollImmediate(time.Second*5, time.Minute*5, func() (bool, error) {
+		klog.V(4).Info(log("probing VolumeAttachment [id=%v]", attachID))
+		attach, err := c.k8s.StorageV1().VolumeAttachments().Get(attachID, meta.GetOptions{})
+		if err != nil {
+			klog.Error(log("attacher.WaitForAttach failed for volume [%s] (will continue to try): %v", volumeHandle, err))
+			return false, fmt.Errorf("volume %v has GET error for volume attachment %v: %v", volumeHandle, attachID, err)
+		}
+		successful, err := verifyAttachmentStatus(attach, volumeHandle)
+		return successful, err
+	})
+	if err != nil {
+		return "", err
+	}
+	return attachID, nil
+}
+
+func verifyAttachmentStatus(attachment *storage.VolumeAttachment, volumeHandle string) (bool, error) {
+	// if being deleted, fail fast
+	if attachment.GetDeletionTimestamp() != nil {
+		klog.Error(log("VolumeAttachment [%s] has deletion timestamp, will not continue to wait for attachment", attachment.Name))
+		return false, errors.New("volume attachment is being deleted")
+	}
+	// attachment OK
+	if attachment.Status.Attached {
+		return true, nil
+	}
+	// driver reports attach error
+	attachErr := attachment.Status.AttachError
+	if attachErr != nil {
+		klog.Error(log("attachment for %v failed: %v", volumeHandle, attachErr.Message))
+		return false, errors.New(attachErr.Message)
+	}
+	return false, nil
+}
+
+func (c *csiAttacher) VolumesAreAttached(specs []*volume.Spec, nodeName types.NodeName) (map[*volume.Spec]bool, error) {
+	klog.V(4).Info(log("probing attachment status for %d volume(s) ", len(specs)))
+
+	attached := make(map[*volume.Spec]bool)
+
+	for _, spec := range specs {
+		if spec == nil {
+			klog.Error(log("attacher.VolumesAreAttached missing volume.Spec"))
+			return nil, errors.New("missing spec")
+		}
+		pvSrc, err := getPVSourceFromSpec(spec)
+		if err != nil {
+			attached[spec] = false
+			klog.Error(log("attacher.VolumesAreAttached failed to get CSIPersistentVolumeSource: %v", err))
+			continue
+		}
+		driverName := pvSrc.Driver
+		volumeHandle := pvSrc.VolumeHandle
+
+		skip, err := c.plugin.skipAttach(driverName)
+		if err != nil {
+			klog.Error(log("Failed to check CSIDriver for %s: %s", driverName, err))
```

Comment:
no need to return here ?

Reply:
Thanks. Already remove it since currently we do not need it.